content/blog/2019-12-18-halide-autoscheduler/index.md

```diff
@@ -0,0 +1,359 @@
+rectangular (as opposed to, say, polytopes in the polyhedral model),
+there is a simple method for doing this:
+we only need to check the maximum and minimum points of the caller functions
+and check the arguments to the callee.
+
+In the example above, the extent of `f` is defined by the box bounded by
+`(1,1)` and `(512, 512)`.
+The arguments to `g` at these points are:
+
+* at `(1,1)`: `(1,1), (2,2)`
+* at `(512,512)`: `(512,512), (513,513)`
+
+Thus we can determine the extent of `g` to be 513x513.
+
+We encode these caller-callee relationships into logical formulas and
+use the [Z3][z3] SMT solver to a retrieve model that contains concrete values
+for the arguments.
+
+[z3]: https://github.com/Z3Prover/z3
+
+
+### Search Algorithm for Schedules
+
+Once loop sizes have been inferred, we have enough information to determine
+important execution features of the schedule, such as how much memory
+it will allocate and how many operations it will perform.
+The *cost* of the schedule is then a weighted sum of these data points.
+
+By default our implementation groups execution features into the following:
+
+* *mem* - amount of memory allocated
+
+* *loads* - number of intermediate results loaded from storage
+
+* *stores* - number of intermediate results stored
+
+* *arithmetic operations* - number of `+`, `-`, `*` and `/` operations performed
+
+* *mathematical operations* - number of `sin`, `cos`, `tan`, `sqrt` operations
+  performed
+
+Each of these groups has a weight that determines the importance of these
+features with respect to the schedule's cost (see **Evaluation** below).
+
+Now that we can give a notion of cost to schedules, we can search for efficient
+schedules.
+We use beam search as our search algorithm, with the default schedule as
+the starting node.
+We describe the concrete parameters used for search below in **Evaluation**.
+
+
+### Conversion to Halide
+
+Once we have a candidate schedule tree, we convert it into Halide.
+We do this by checking the ancestry path from compute nodes:
+this path determines whether a function's variables are split,
+the traversal order for computing the function, and, for called functions,
+the granularity at which the function is stored and computed.
+
+Consider the schedule above for the functions `bx`, `by`, and `in`.
+Converted into Halide code, the schedule looks like the following:
+
+```
+by.reorder(y, x);
+
+bx.store_root();
+bx.compute_at(by, y);
+bx.reorder(y, x);
+```
+
+
+## Evaluation
+
+We evaluate the performance of the autoscheduler over three benchmarks.
+We do this by comparing the performance of the autoscheduled run
+(OPT configuration) vs. the run with the default schedule (DEF configuration).
+We measure runtime and memory usage using `gprof`.
+
+For the experiments, we set the weights for execution features as follows:
+
+Group       | Weight 
+------------|---------------
+mem         | 0.1
+loads       | 0.5
+stores      | 0.5
+arith ops   | 1.0
```

Comment:
Interesting! As long as we're guessing, I would have guessed load & store costs to be _more_ than that for arithmetic ops. I guess it depends heavily on which level of the memory hierarchy the data ends up fitting into…